### ISSUE TITLE
Passing parameters to helper in Meteor's dynamic templates

### ISSUE THREAD

**Issue by cneuro** [meteor/meteor#4303]
I realise that this has been brought up before, and that `Template.dynamic` isn't designed to take in a parameter if its `template` parameter is a helper, but there is a use case which can't be accomplished right now. This is when the dynamic template's name depends on more than just `Session` variables and data contexts.

Here is what I would like to do:

```javascript
// a global helper that composites the template's name using domain-specific and global parameters
Template.registerHelper('templateName', function (name) {
    return name + Session.get('someVariable');
});
```

```html
<!-- use case: a template calling two dynamic ones -->
<template name="someTemplate">
  <div class="some-class">
    {{> Template.dynamic template=templateName 'title' }}
  </div>
  <div class="another-class">
    {{> Template.dynamic template=templateName 'content' }}
  </div>
</template>
```

This pattern is extremely DRY and it avoids having to set up nested conditionals and rewrite quasi-identical templates each with minimal changes.

Right now, I've got this:

```javascript
Template.registerHelper('templateName', function () {
  var dt = this.dName || Template.parentData().dName;
  return dName + Session.get('someVariable'));
});
```

```html
<template name="someTemplate">
  {{> segment dName="title"}}
  {{> segment dName="content"}}
</template>

<template name="segment">
   <div class="some-class">
    {{> Template.dynamic template=templateName }}
   </div>
</template>
```

It works, but it isn't ideal, because;

- confusion-prone need to include the parameter for the dynamic template's name in the parent template's call
- only ever being allowed one `Template.dynamic` per template due to one parameter, leading to scalability issues
- putting the `dName` parameter in the template's data context is mixed in with local data, requiring the hacky check whether it's accessible in the current one or the parent's

Are there plans to add this functionality? Am I going about this the wrong way? Did anyone else run into these issues?

Thanks for reading.

**Comment by benjamn** [comment]
I can see the appeal of what you're asking for, but `Template.dynamic` also accepts a `data=` keyword argument for specifying the data context, which might be enough for your purposes? That is, `{{> Template.dynamic template=templateName data="title" }}`.

**Comment by calvin-evans** [comment]
This is just a +1 really but I'm regularly in a similar situation to @cneuro. 

Right now I have 3 possible status's for an item which require different templates. Problem is, the names are too generic to use as a template name (won, lost, pending). So I have two choices really. Create an additional `templateName` field in my collection or do my processing before using `Template.dynamic` like @cneuro. Both of which are less than ideal

As he's already said, would be great to be able to do something like this:

`{{> Template.dynamic template="getTemplate status}}`

**Comment by cneuro** [comment]
@benjamn, I already pass a named `data` attribute to the template with the dynamic one inside it, which is a bit more complex than what you're suggesting, but more scalable, because you only need to write one dynamic template name helper. Ultimately, as @calvin-evans is also suggesting, these approaches are only really viable for small, manual delegation of dynamic template rendering calls. I'm thinking more of an abstracted, concise, app-wide way to call dynamic templates with the custom parameters inside the call itself.

I came across another way to do it from this [Meteor forum post](https://forums.meteor.com/t/pass-argument-to-helper-in-template-dynamic-call/3971), which suggests wrapping a `{{#with}}` context around the call to `{{Template.dynamic}}`. This is probably more concise than what I'm doing right now in terms of sending a custom data string into a parent template.

**Comment by avital** [comment]
Do spacebars nested expressions help? (https://github.com/meteor/meteor/pull/4101)

You could then do something like `{{> Template.dynamic template=(getTemplate "foo") data=data}}` which would call the `getTemplate` helper with the argument `"foo"` to decide which template to load

**Comment by calvin-evans** [comment]
@avital Perfect. Guess I'll wait for the next release :+1:

**Comment by cneuro** [comment]
Spacebars nested expressions is exactly the solution! Looks like it's also now part of the devel branch, but I can't tell if that's merged with the current release 1.1.0.2 (it throws a Type Error).

@calvin-evans, yep looks like we just need to wait :)

**Comment by calvin-evans** [comment]
@cneuro yeah still getting this error in 1.1.0.2 https://github.com/meteor/meteor/commit/a46aa21a010ea56419a452c396640d60c72f7a3a#diff-a73cc28f54100fa2d786662303f632d3L224
